numpy/_core/src/umath/loops.c.src

```diff
@@ -471,13 +471,48 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 int
 }
 /**end repeat1**/
 
+static inline @type@ _@TYPE@_squared_exponentiation_helper(@type@ base, @type@ exponent_two, int first_bit) {
+   // Helper method to calculate power using squared exponentiation
+   // The algorithm is partly unrolled. The second and third argument are the exponent//2 and the first bit of the exponent
+   @type@ out = first_bit ? base : 1;
+   while (exponent_two > 0) {
+        base *= base;
+        if (exponent_two & 1) {
+            out *= base;
+        }
+        exponent_two >>= 1;
+    }
+   return out;
+}
+
 NPY_NO_EXPORT void
 @TYPE@_power(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
 {
+    if (steps[1]==0) {
+        // stride for second argument is 0
+        BINARY_DEFS_ZERO_STRIDE
+        const @type@ in2 = *(@type@ *)ip2;
+        #if @SIGNED@
+                if (in2 < 0) {
+                    npy_gil_error(PyExc_ValueError,
+                                  "Integers to negative integer powers are not allowed.");
+                    return;
+                }
+        #endif
+
+        int first_bit = in2 & 1;
+        @type@ in2start = in2 >> 1;
+
+        BINARY_LOOP_SLIDING_ZERO_STRIDE {
+            @type@ in1 = *(@type@ *)ip1;
+
+            *((@type@ *) op1) = _@TYPE@_squared_exponentiation_helper(in1, in2start, first_bit);
+        }
+        return;
+    }
     BINARY_LOOP {
```

Comment:
I am happy to keep it this way, but I was half hoping that we can put this whole `BINARY_LOOP {` into a `static inline` helper and then just write:
```
if (steps[1] == 0) {
    @TYPE@_power_impl(args, dimensions, steps);
else {
    @TYPE@_power_impl(args, dimensions, steps);
}
```
Unfortunately, I suspect that the compiler cannot lift the `npy_gil_error` out of the loop so not sure it is worthwhile.
(The function could have `if (scalar_pow)` which is passed in and the compiler should optimize it away, though.)